presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,109 @@ private PinotExpression handleLogicalBinary(
+            // Trying to pushdown the predicate of comparing a time column to a time literal.
+            return handleDateOrTimestampBinaryExpression(operator, arguments, context).orElseGet(
+                    () -> derived(format(
+                            "(%s %s %s)",
+                            arguments.get(0).accept(this, context).getDefinition(),
+                            operator,
+                            arguments.get(1).accept(this, context).getDefinition())));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private Optional<PinotExpression> handleDateOrTimestampBinaryExpression(String operator, List<RowExpression> arguments, Function<VariableReferenceExpression, Selection> context)
+    {
+        // Handle the binary comparison logic of <DATE/TIMESTAMP field> <binary op> <DATE/TIMESTAMP literal>.
+        // Pinot stores time as:
+        //   - `DATE`: Stored as `INT`/`LONG` `daysSinceEpoch` value
+        //   - `TIMESTAMP`: Stored as `LONG` `millisSinceEpoch` value.
+        // In order to push down this predicate, we need to convert the literal to the type of Pinot time field.
+        // Below code compares the time type of both side:
+        //   - if same, then directly push down.
+        //   - if not same, then convert the literal time type to the field time type.
+        // Take an example of comparing a `DATE` field to a `TIMESTAMP` literal:
+        //   - Sample predicate: `WHERE eventDate < current_time`.
+        //   - Input type is the `eventDate` field data type, which is `DATE`.
+        //   - Expect type is the right side `literal type`, which means right side is `TIMESTAMP`.
+        // The code below converts `current_time` from `millisSinceEpoch` value to `daysSinceEpoch` value, which is
+        // comparable to values in `eventDate` column.
+        String leftDefinition = arguments.get(0).accept(this, context).getDefinition();
+        String rightDefinition = arguments.get(1).accept(this, context).getDefinition();
+        Optional<String> left = handleTimeValueCast(arguments.get(1), arguments.get(0), leftDefinition);
+        Optional<String> right = handleTimeValueCast(arguments.get(0), arguments.get(1), rightDefinition);
+        if (left.isPresent() && right.isPresent()) {
+            return Optional.of(derived(format("(%s %s %s)", left.get(), operator, right.get())));
+        }
+        else {
+            return Optional.empty();
+        }
+    }
+
+    private boolean isDateTimeConstantType(Type type)
+    {
+        return type == DateType.DATE || type == TimestampType.TIMESTAMP || type == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+    }
+
+    private Optional<String> handleTimeValueCast(RowExpression timeFieldExpression, RowExpression timeValueExpression, String timeValueString)
+    {
+        Type inputType;
```

Comment:
I think that giant block of comment makes more sense for this method. Perhaps also amend that block to inform what the Optional.empty return signifies

Reply:
done